pyam_analysis/core.py

```diff
@@ -65,71 +65,94 @@ class IamDataFrame(object):
     (including validation of values, completeness of variables provided)
     as well as a number of visualization and plotting tools."""
 
-    def __init__(self, data, regions=None, **kwargs):
+    def __init__(self, data, regions=None, variables=None, units=None,
+                 years=None, **kwargs):
         """Initialize an instance of an IamDataFrame
 
         Parameters
         ----------
-        data: pyam-analyis.IamDataFrame, ixmp.TimeSeries, ixmp.Scenario
-        or data file
-            an instance of an IamDataFrame, TimeSeries or Scenario
-            (these two option require the 'ixmp' package as a dependency)
-            or a file with IAMC-style data
-        regions: list
-            list of regions to be imported
+        data: ixmp.TimeSeries, ixmp.Scenario, pandas.DataFrame or data file
+            an instance of an TimeSeries or Scenario (requires `ixmp`),
+            or pandas.DataFrame or data file with IAMC-format data columns
+        regions : list of strings
+            filter by regions
+        variables : list of strings
+            filter by variables (only with ixmp objects)
+        units : list of strings
+            filter by units (only with ixmp objects)
+        years : list of integers
+            filter by years (only with ixmp objects)
         """
-        # copy-constructor
-        if isinstance(data, IamDataFrame):
-            self.data = data.data
-            self._meta = data._meta
-            self.cat_color = data.cat_color
-            self.col_count = data.col_count
-
-        # read data from source
+        # import data from pandas.DataFrame or read from source
+        if isinstance(data, pd.DataFrame):
+            self.data = format_data(data)
+        elif has_ix and isinstance(data, ixmp.TimeSeries):
+            self.data = read_ix(data, regions=regions, variables=variables,
+                                units=units, years=years)
         else:
-            if has_ix and isinstance(data, ixmp.TimeSeries):
-                self.data = read_ix(data, regions)
-            else:
-                self.data = read_data(data, regions, **kwargs)
+            self.data = read_data(data, regions, **kwargs)
 
-            # define a dataframe for categorization and other meta-data
-            self._meta = return_index(self.data, mod_scen,
-                                      drop_duplicates=True)
-            self.reset_category(True)
+        # define a dataframe for categorization and other meta-data
+        self._meta = return_index(self.data, mod_scen,
+                                  drop_duplicates=True)
+        self.reset_category(True)
```

Comment:
use explicit argument name (`reset_exclude=True`)

Reply:
was on my to-do list as part of the category clean-up...

Reply:
Sounds good.